### DISCUSSION TITLE
Input dataframe for statsforecast

### DISCUSSION THREAD

**Discussion by AFilipaVEL** [Nixtla/statsforecast#779] [Q&A] [answered]
Hi all,

I'm new to Python, PySpark and StatsForecast... i'm now trying to run a simple forecast example to get familiar with this module. My first step was to create a dataframe with the mandatory columns `unique_id` (string), `ds` (date yyyy-mm-dd) and `y` (float). 

I converted it  to a pyspark.pandas.frame.DataFrame and it looks good. However, when i apply the function `StatsForecast.plot(pd_df)` (being `pd_df` my pandas dataframe), i get the following error:

> AttributeError: 'NoneType' object has no attribute 'Date'

Function `StatsForecast` and `forecast` method work ok.

Do you have any idea of what might be happening here?

Thanks in advance!

**Comment by jmoralez** [comment]
Hey @AFilipaVEL, thanks for using statsforecast. Can you provide the code that you're running or the full stacktrace of the error?

**Reply by AFilipaVEL** [reply]
Hi @jmoralez thank you, for developing this great tool!

Here's the code i'm using, knowing that `st` is my Spark dataframe:

```
import pandas as pd

from statsforecast import StatsForecast
from statsforecast.models import AutoARIMA, AutoETS

import os
os.environ['NIXTLA_ID_AS_COL'] = '1'

pd_st = st.pandas_api()

StatsForecast.plot(pd_st)
```

The full error is


```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
File <command-4317370202810021>, line 1
----> 1 StatsForecast.plot(pd_st)

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-6cb1a290-e45a-42ad-950f-391093b99775/lib/python3.10/site-packages/statsforecast/core.py:1314, in _StatsForecast.plot(df, forecasts_df, unique_ids, plot_random, models, level, max_insample_length, plot_anomalies, engine, id_col, time_col, target_col, resampler_kwargs)
   1282 """Plot forecasts and insample values.
   1283 
   1284 Parameters
   (...)
   1310     its `show_dash` method.
   1311 """
   1312 from utilsforecast.plotting import plot_series
-> 1314 df = ensure_time_dtype(df, time_col)
   1315 if isinstance(df, pd.DataFrame) and df.index.name == id_col:
   1316     warnings.warn(
   1317         "Passing the ids as the index is deprecated. "
   1318         "Please provide them as a column instead.",
   1319         category=FutureWarning,
   1320     )

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-6cb1a290-e45a-42ad-950f-391093b99775/lib/python3.10/site-packages/utilsforecast/validation.py:51, in ensure_time_dtype(df, time_col)
     48 """Make sure that `time_col` contains timestamps or integers.
     49 If it contains strings, try to cast them as timestamps."""
     50 times = df[time_col]
---> 51 if _is_dt_or_int(times):
     52     return df
     53 parse_err_msg = (
     54     f"Failed to parse '{time_col}' from string to datetime. "
     55     "Please make sure that it contains valid timestamps or integers."
     56 )

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-6cb1a290-e45a-42ad-950f-391093b99775/lib/python3.10/site-packages/utilsforecast/validation.py:35, in _is_dt_or_int(s)
     34 def _is_dt_or_int(s: Series) -> bool:
---> 35     return _is_dt_dtype(s) or _is_int_dtype(s)

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-6cb1a290-e45a-42ad-950f-391093b99775/lib/python3.10/site-packages/utilsforecast/validation.py:30, in _is_dt_dtype(s)
     28     out = pd.api.types.is_datetime64_any_dtype(s.dtype)
     29 else:
---> 30     out = s.dtype in (pl.Date, pl.Datetime)
     31 return out

AttributeError: 'NoneType' object has no attribute 'Date'

```

**Reply by jmoralez** [reply]
The `pandas_api` method doesn't return a pandas dataframe, but a `PandasOnSparkDataFrame`. Can you try using `st.toPandas()` instead? Just be careful to pull something of reasonable size, you may want to filter some ids first, for example: `pd_st = st.where(col('unique_id').isin(['id0', 'id1'])).toPandas()`

**Reply by AFilipaVEL** [reply]
It worked! Muchísimas gracias!!

Nevertheless, i'm starting with a very small dataframe just to learn well the methods but my intention is to provide a huge dataframe with millions of rows. Will this approach work?

**Reply by jmoralez** [reply]
It will work for the `forecast` method, which will also return a distributed dataframe. For plotting you should always try to filter first and then convert to pandas to visualize a small subset of the series.

**Reply by AFilipaVEL** [reply]
Well, i won't really need to plot, this is just to "walk through" the module. Do you recommend using Fugue?

**Reply by jmoralez** [reply]
We use fugue for the implementation but you shouldn't have to think about it, when you provide a spark dataframe to the forecast method you'll get a spark dataframe as output.

**Reply by AFilipaVEL** [reply]
I apologize but now i'm confused. I read that `statsForecast` should receive a Pandas dataframe and that Fugue would "communicate" with Spark to speed things up. Sorry the not-so-technical terms...

Now you're saying that `statsForecast` is fast enough if it receives a spark dataframe? Even when there are millions of time series to model and predict? 

If this is true, should i use `n_jobs = -1` in `statsForecast ` to take full advantages of my cluster?

**Reply by jmoralez** [reply]
It automatically dispatches to the corresponding backend:
* If you provide a pandas dataframe it will run locally. Here you can set n_jobs>1 to use multiprocessing in your local machine.
* If you provide a spark dataframe it will run in your cluster, here n_jobs will automatically be set to 1 to avoid nested parallelism (spark handles the parallelism here). To maximize the utilization of your cluster you should make sure that you at least have as many partitions as executors (or a multiple of them) so that spark can schedule at least one task per executor.

**Reply by AFilipaVEL** [reply]
Ok, great! So simple!

Another thing, if i'm allowed... if parallelism is ensured by providing a spark dataframe, then i assume i don't need module `statsforecast.distributed.utils` ? I tried to run 

```from statsforecast.distributed.utils import forecast``` 

 but i get the error

```
ModuleNotFoundError: No module named 'statsforecast.distributed.utils'
```

**Reply by jmoralez** [reply, marked as answer]
No, you don't need it. Just provide a spark dataframe to `StatsForecast.forecast`

**Reply by AFilipaVEL** [reply]
Wow! Many, many thanks! This was really useful!! I wish you all the best and keep up the good work!
